### PR TITLE
GH-4814 Looking at Gulp/Jake/Grunt plugins extensions.

### LINKED ISSUE #4814: Load built-in vscode plugins to Theia: grunt, gulp, jake
### Description
Capability to load built-in vscode plugins to Theia
Looking at grunt, gulp and jake

### Reproduction Steps
Put the jake, grunt and gulp vscode built-in plugins into the theia/plugins folder
unzip/untar the test project attached
npm install  // to build the node_modules
Start theia and load the test project into the browser
Note:
1- Load the test program (Note, just to test that we can extract the task as "GULP, GRUNT, JAKE, NPM"
2- For Jake, still under investigation

**OS and Theia version:**
Ubuntu 16.04
Theia 0.5.0

**Diagnostics:** 

[test.tar.gz](https://github.com/theia-ide/theia/files/3045123/test.tar.gz)

### PR DESCRIPTION
**Pull request by lmcbout**
Fixes #4814 
It is possible to have tasks with the same type, so we need to
differentiate them.
Using the handle available with the task to create a unique provider.
This allow to have following plugins within the same workspace:
- grunt
- gulp
- jake

Jake is using the old API to computeTask(). Because of that, I need to implement the deprecated Task constructor to handle Jake. 
To support the vscode built-in plugin for JAKE , there is two ways:
1-   I need to implement the deprecated Task constructor to handle Jake. 
2-  Jake vscode built-in plugin needs to be adjusted to handle the creation of task using the newest API

In the current solution, I implemented solution 1, so we can support JAKE until vscode modify the Jake plugin

For solution 2, we need to modify:
File:  jake/src/main.ts
Method computeTasks() at ~line 145, need to modify the following line by adding the workspaceFolder as a parameter:

OLD:
let task = new vscode.Task(kind, taskName, 'jake', new vscode.ShellExecution(`${jakeCommand} ${taskName}`, options));

 NEW:
let task = new vscode.Task(kind, **this._workspaceFolder,** taskName, 'jake', new vscode.ShellExecution(`${jakeCommand} ${taskName}`, options));

For testing purpose, you need to add in the package.json the following packages. (Either in devDependencies or dependencies)
"dependencies": {
  "gulp": "latest",
     "gulp-util": "latest",
     "gulp-sass": "latest",
     "gulp-coffee": "latest",
     "gulp-uglify": "latest",
     "gulp-concat": "latest",
     "gulp-connect": "latest",
  "grunt": "~0.4.5",
     "grunt-cli": "^1.3.2",
     "grunt-contrib-jshint": "~0.6.4",
     "grunt-contrib-nodeunit": "~0.2.0",
     "grunt-contrib-watch": "~0.5.3",
  "jake": "^8.1.1"    
}
With those packages installed, you can test "Gulp", Grunt", "Jake" and "npm" if installed

Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

File: packages/task/src/browser/task-contribution.ts [hunk 1]
```diff
@@ -70,15 +70,22 @@ export class TaskProviderRegistry {
...
+        this.providers.set(key, provider);
         return {
-            dispose: () => this.providers.delete(type)
+            dispose: () => this.providers.delete(key)
         };
     }
 
-    getProvider(type: string): TaskProvider | undefined {
-        return this.providers.get(type);
+    getProvider(type: string): TaskProvider[] {
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-contribution.ts]: Does the return type not break API?
If anyone called `getProvider` they expect either a `TaskProvider` or `undefined` but now we return an array.
I assume that is why we have the method `getProviders` which already returns an array.

Comment (lmcbout) [reply packages/task/src/browser/task-contribution.ts]: @vince-fugnitto Thanks for the review, code adjusted and those providers return to their original values.

File: packages/plugin-ext/src/plugin/types-impl.ts [hunk 2]
```diff
@@ -1541,7 +1541,7 @@ export enum TaskScope {
 
 export class Task {
     private taskDefinition: theia.TaskDefinition | undefined;
-    private taskScope: theia.TaskScope.Global | theia.TaskScope.Workspace | theia.WorkspaceFolder | undefined;
+    private taskScope?: theia.TaskScope.Global | theia.TaskScope.Workspace | theia.WorkspaceFolder | undefined;
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/plugin/types-impl.ts]: why is the `?` needed, `undefined` is already supported.

Comment (lmcbout) [reply packages/plugin-ext/src/plugin/types-impl.ts]: Removed

**Comment by elaihau** [comment]
this pr is part of https://github.com/theia-ide/theia/issues/4216

**Pull request merged**